File: array/121/121.cpp
/*
 * @Description: 121 买卖股票的最佳时机
 * @Version: 1.0
 * @Autor: laijiatao
 * @Date: 2021-07-25 10:15:00
 * @LastEditors: laijiatao
 * @LastEditTime: 2021-07-25 10:40:17
 */

/**
 * @description: 第一次做，超时
 * @param {*}
 * @return {*}
 * @author: laijiatao
 */
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if(prices.size() <= 1) {
            return 0;
        }

        int max = 0;
        int liRui = INT_MIN;
        for(int i = 0; i < prices.size(); i++) {
            for(int j = i + 1; j < prices.size(); j++) {
                liRui = prices[j] - prices[i];
                if(max < liRui) {
                    max = liRui;
                }
            }
        }

        return max;
    }
};

/**
 * @description: 双指针： https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock/solution/shi-yong-shuang-zhi-zhen-ke-jie-by-chi-h-xd9o/
 * @param {*}
 * @return {*}
 * @author: laijiatao
 */

class Solution {
    public int maxProfit(int[] prices) {
        int left = 0, right = 0;
        int idax = 0;
        while ( right != prices.length ){
            idax = Math.max(prices[right] - prices[left], idax);
            if ( prices[right] < prices[left] )left = right;
            right++;
        }return idax;
    }
}

/**
 * @description: 动态规划：https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock/solution/gu-piao-wen-ti-python3-c-by-z1m/
 * @param {*}
 * @return {*}
 * @author: laijiatao
 */
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        if n == 0: return 0 # 边界条件
        dp = [0] * n
        minprice = prices[0] 

        for i in range(1, n):
            minprice = min(minprice, prices[i])
            dp[i] = max(dp[i - 1], prices[i] - minprice)

        return dp[-1]

File: array/121/121.py
'''
给定一个数组 prices ，它的第 i 个元素 prices[i] 表示一支给定股票第 i 天的价格。

你只能选择 某一天 买入这只股票，并选择在 未来的某一个不同的日子 卖出该股票。设计一个算法来计算你所能获取的最大利润。

返回你可以从这笔交易中获取的最大利润。如果你不能获取任何利润，返回 0 。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。

1 dp数组以及下标的含义
dp[i] 前i天的的最大收益
2 递推公式
统计出前i天的最低价格，minPrice
dp[i] = max(dp[i-1], price[i]-minPrice)
3 dp数组初始化
minPrice = price[0]
dp[i] = 0
4 遍历顺序
从前往后
5 打印dp数组
'''

class Solution:
    def maxProfit(self, prices):
        if len(prices) == 0:
            return 0
        dp = [0] * len(prices)
        minPrice = prices[0]

        for i in range(1, len(prices)):
            minPrice = min(minPrice, prices[i-1])
            dp[i] = max(dp[i-1], prices[i]-minPrice)
        
        print(dp)
        return dp[-1]

if __name__ == '__main__':
    s = Solution()
    print(s.maxProfit([7,1,5,3,6,4]))

